src/sys/fuchsia/eventedfd.rs

```diff
@@ -0,0 +1,246 @@
+// happen on `Drop`. These accesses are safe because `drop` can only be called at most once from
+// a single thread, and after it is called no other functions can be called on the `EventedFdInner`.
+unsafe impl Sync for EventedFdInner {}
+unsafe impl Send for EventedFdInner {}
+
+#[derive(Clone, Debug)]
+pub struct EventedFd {
+    pub inner: Arc<EventedFdInner>
+}
+
+impl EventedFd {
+    pub unsafe fn new(fd: RawFd) -> Self {
+        let mxio = sys::__mxio_fd_to_io(fd);
+        assert!(mxio != ::std::ptr::null(), "FileDescriptor given to EventedFd must be valid.");
+
+        EventedFd {
+            inner: Arc::new(EventedFdInner {
+                registration: Mutex::new(None),
+                fd: fd,
+                mxio: mxio,
+            })
+        }
+    }
+}
+
+impl Evented for EventedFd {
+    fn register(&self,
+                poll: &Poll,
+                token: Token,
+                interest: Ready,
+                opts: PollOpt) -> io::Result<()>
+    {
+        let epoll_events = ioevent_to_epoll(interest, opts);
+
+        let (handle, raw_handle, signals) = unsafe {
+            let mut raw_handle: sys::mx_handle_t = mem::uninitialized();
+            let mut signals: sys::mx_signals_t = mem::uninitialized();
+            sys::__mxio_wait_begin(self.inner.mxio, epoll_events, &mut raw_handle, &mut signals);
+
+            // We don't have ownership of the handle, so we can't drop it
+            let handle = DontDrop::new(magenta::Handle::from_raw(raw_handle));
+            (handle, raw_handle, signals)
+        };
+
+
+        let needs_rereg = opts.is_level() && !opts.is_oneshot();
+
+        {
+            let mut registration_lock = self.inner.registration.lock().unwrap();
+            if registration_lock.is_some() {
+                return Err(io::Error::new(
+                    io::ErrorKind::AlreadyExists,
+                    "Called register on an already registered file descriptor."));
+            }
+            *registration_lock = Some(EventedFdRegistration {
+                token: token,
+                handle: DontDrop::new(unsafe { magenta::Handle::from_raw(raw_handle) }),
+                rereg_signals: if needs_rereg {
+                    Some((signals, poll_opts_to_wait_async(opts)))
+                } else {
+                    None
+                },
+            })
+        }
+
+        let registered = poll::selector(poll)
+            .register_fd(handle.inner_ref(), self, token, signals, opts);
+
+        if registered.is_err() {
+            let mut registration_lock = self.inner.registration.lock().unwrap();
+            *registration_lock = None;
+        }
+
+        registered
+    }
+
+    fn reregister(&self,
+                  poll: &Poll,
+                  token: Token,
+                  interest: Ready,
+                  opts: PollOpt) -> io::Result<()>
+    {
+        // Take out the registration lock
```

Comment:
Could this be a short explanation of why the lock needs to be taken out across both deregistration and reregistration? Right now it's kind of redundant with the code. 

I don't understand `mio` enough to know why this implementation of `reregister` takes out the lock once for both the deregistration and the reregistration, instead of the entire implementation just being `self.deregister(...);self.register(...)`. 

Is it because a concurrent call to `deregister` between those two would crash in `reregister` instead of the concurrent `register`? Or because a concurrent call to `deregister` in the middle would crash when it wouldn't afterwards (although this is a race condition because if it is scheduled before then `reregister` will crash)? Is it just a performance optimization?